src/generic_inverse.cpp

```diff
@@ -0,0 +1,115 @@
+/******************************************************************************
+ *
+ * Project:  PROJ
+ * Purpose:  Generic method to compute inverse projection from forward method
+ * Author:   Even Rouault <even dot rouault at spatialys dot com>
+ *
+ ******************************************************************************
+ * Copyright (c) 2018, Even Rouault <even dot rouault at spatialys dot com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "proj_internal.h"
+
+#include <algorithm>
+#include <cmath>
+
+/** Compute (lam, phi) corresponding to input (xy.x, xy.y) for projection P.
+ *
+ * Uses Newton-Raphson method, extended to 2D variables, that is using
+ * inversion of the Jacobian 2D matrix of partial derivatives. The derivatives
+ * are estimated numerically from the P->fwd method evaluated at close points.
+ *
+ * Note: thresholds used have been verified to work with adams_ws2 and wink2
+ *
+ * Starts with initial guess provided by user in lpInitial
+ */
+PJ_LP pj_generic_inverse_2D_with_newton_raphson(PJ_XY xy, PJ *P,
```

Comment:
The name is quite long, since there's only one generic inverse function around wouldn't it be fine to simply call it `pj_generic_inverse_2d`?

Reply:
done

Reply:
Thanks. Pedantic follow-up: The name should also be changed in the commit message which now states:

> - Move the generic method initiated from adams_ws2 to a
>   pj_generic_inverse_2D_with_newton_raphson() method
> - Use it in adams_ws2
> - Use it in wink2

Reply:
ah right. done

Reply:
We need the ability to suggest changes to commit messages as well :-) Just stumpled on this one by coincidence - figured it may as well be updated.